protocol_decoder/src/decoding.rs

```diff
@@ -264,26 +279,107 @@ impl ProcessedBlockTrace {
         Ok(())
     }
 
+    /// Pads a generated IR vec with additional "dummy" entries if needed.
+    /// We need to ensure that generated IR always has at least `2` elements,
+    /// and if there are only `0` or `1` elements, then we need to pad so
+    /// that we have two entries in total. These dummy entries serve only to
+    /// allow the proof generation process to finish. Specifically, we need
+    /// at least two entries to generate an agg proof, and we need an agg
+    /// proof to generate a block proof. These entries do not mutate state
+    /// (unless there are withdrawals in the block (see
+    /// `[add_withdrawals_to_txns]`), where the final one will mutate the
+    /// state trie.
     fn pad_gen_inputs_with_dummy_inputs_if_needed(
         gen_inputs: &mut Vec<TxnProofGenIR>,
         other_data: &OtherBlockData,
         initial_trie_state: &PartialTrieState,
-    ) {
+    ) -> bool {
         match gen_inputs.len() {
             0 => {
-                // Need to pad with two dummy txns.
+                // Need to pad with two dummy entries.
                 gen_inputs.extend(create_dummy_txn_pair_for_empty_block(
                     other_data,
                     initial_trie_state,
                 ));
+
+                true
             }
             1 => {
+                // Just need one.
                 let dummy_txn = create_dummy_gen_input(other_data, initial_trie_state, 0);
                 gen_inputs.insert(0, dummy_txn);
+
+                true
             }
-            _ => (),
+            _ => false,
         }
     }
+
+    /// The withdrawals are always in the final ir payload. How they are placed
+    /// differs based on whether or not there are already dummy proofs present
+    /// in the IR. The rules for adding withdrawals to the IR list are:
+    /// - If dummy proofs are already present, then the withdrawals are added to
+    ///   the last dummy proof (always index `1`).
+    /// - If no dummy proofs are already present, then a dummy proof that just
+    ///   contains the withdrawals is appended to the end of the IR vec.
+    fn add_withdrawals_to_txns(
+        txn_ir: &mut Vec<TxnProofGenIR>,
+        other_data: &OtherBlockData,
+        final_trie_state: &mut PartialTrieState,
+        withdrawals: Vec<(Address, U256)>,
+        dummies_already_added: bool,
+    ) -> TraceParsingResult<()> {
+        // Withdrawals update balances in the account trie, so we need to do that here.
+        Self::update_trie_state_from_withdrawals(&withdrawals, &mut final_trie_state.state)?;
+
+        match dummies_already_added {
+            false => {
+                // Guaranteed to have a real txn.
+                let final_ir = txn_ir.last().unwrap();
+
+                // Dummy state will be the state after the final txn.
+                let withdrawal_dummy =
+                    create_dummy_gen_input(other_data, final_trie_state, final_ir.txn_idx + 1);
```

Comment:
same here, we are having identical state for before and after for this dummy. While it should be 

initial: final_trie_state
// update final_trie_state with withdrawals
final: final_trie_state